llmfoundry/callbacks/curriculum_learning_callback.py

```diff
@@ -0,0 +1,103 @@
+        # Check if we are using a DataLoader and StreamingDataset
+        if not isinstance(dataloader, DataLoader):
+            raise ValueError(
+                f'CurriculumLearning callback can only be used with a train ',
+                f'dataloader of type DataLoader, but got {type(dataloader)}.')
+        dataset = dataloader.dataset
+        if not isinstance(dataset, StreamingDataset):
+            raise ValueError(
+                f'CurriculumLearning callback only supports StreamingDataset ',
+                f'because it requires loading and saving dataset state. ',
+                f'Instead, got a dataset of type {type(dataset)}')
+        # Save the current dataset state so we can restore it if needed.
+        self.current_dataset_state = dataset.state_dict(0, False)
+
+    def after_load(self, state: State, logger: Logger):
+        del logger
+
+        # As saved_dataset_index is loaded from state_dict, this only run when
+        # a user explicitly increments the dataset_index and not on any other
+        # resumption, including autoresume.
+        train_loader = state._train_dataloader
+        assert isinstance(
+            train_loader,
+            DataLoader), 'CurriculumLearning callback requires a DataLoader.'
+        dataset = train_loader.dataset
+        assert isinstance(
+            dataset, StreamingDataset
+        ), 'CurriculumLearning callback requires a StreamingDataset.'
+        if self.saved_dataset_index < self.dataset_index:
+            # Ignore the dataset state that was read in from the checkpoint, and
+            # replace with the new dataset state. This preserves resumption info.
+            if self.current_dataset_state['epoch'] < 0:
+                # Make sure the epoch in the loaded state dict is not negative.
+                # Otherwise, the state will be considered stale, and will be ignored.
+                self.current_dataset_state['epoch'] = 0
+            dataset.load_state_dict(self.current_dataset_state)
+            # Start a new epoch since we are using a new dataset.
+            # This will also reset the sample_in_epoch written to checkpoint,
+            # making sure that subsequent resumptions proceed correctly.
+            state.timestamp = state.timestamp.to_next_epoch()
+            self.new_dataset_setup = True
+        elif self.dataset_index == 0 and len(self.all_dataset_configs) == 0:
+            # Make sure to save our current dataset config if we are just starting training.
+            self.new_dataset_setup = True
+
+    def state_dict(self):
+        if self.new_dataset_setup and not self.dataset_config_appended:
+            # Append the new dataset config to the list of all dataset configs.
+            self.all_dataset_configs.append(self.current_dataset_config)
+            # Only append the dataset config once, not on every single checkpoint save.
+            self.dataset_config_appended = True
```

Comment:
Can we get rid of the vars here and instead do directly in after_load?

Reply:
modified to do this